Bottom: 22e3b80e1c1bc2fd08f8b277615eb4de224b84af
Top:    8df18e258b71d76573609f28298400e9c9cd6a00
Author: Alexandre Oliva <aoliva@redhat.com>
Date:   2016-03-02 01:14:51 -0300

introduce callback to get type char

---

diff --git a/include/gcc-cp-fe.def b/include/gcc-cp-fe.def
index 4a6fb7f..f0dadc1 100644
--- a/include/gcc-cp-fe.def
+++ b/include/gcc-cp-fe.def
@@ -321,7 +321,8 @@ GCC_METHOD4 (gcc_type, build_method_type,
 
 // FIXME: add default expressions for function arguments?
 
-/* Return an integer type with the given properties.  */
+/* Return an integer type with the given properties.
+   Pass zero in IS_UNSIGNED and SIZE_IN_BYTES to get type 'char'.  */
 
 GCC_METHOD2 (gcc_type, int_type,
 	     int /* bool */,		   /* Argument IS_UNSIGNED.  */
diff --git a/libcc1/libcp1plugin.cc b/libcc1/libcp1plugin.cc
index 7032ac0..9045b7d 100644
--- a/libcc1/libcp1plugin.cc
+++ b/libcc1/libcp1plugin.cc
@@ -1250,8 +1250,13 @@ gcc_type
 plugin_int_type (cc1_plugin::connection *self,
 		 int is_unsigned, unsigned long size_in_bytes)
 {
-  tree result = c_common_type_for_size (BITS_PER_UNIT * size_in_bytes,
-					is_unsigned);
+  tree result;
+
+  if (!size_in_bytes && !is_unsigned)
+    result = char_type_node;
+  else
+    result = c_common_type_for_size (BITS_PER_UNIT * size_in_bytes,
+				     is_unsigned);
   if (result == NULL_TREE)
     result = error_mark_node;
   else
